### ISSUE TITLE
ExactArgs + ValidArgs + OnlyValidArgs

### ISSUE THREAD

**Issue by patrick-motard** [spf13/cobra#745]
I have a command called `i3` that must have at exactly one argument given to it, and the argument can either be "load" or "disable". By passing the command "ExactArgs(1)" I can make it do exactly one arg, but it doesn't respect "ValidArgs". I can pass "OnlyValidArgs" to make it respect "ValidArgs", but I can't pass it both. Something like `Args: cobra.ExactArgs(1) && cobra.OnlyValidArgs(cobra, args)` would be awesome.

```
var i3Cmd = &cobra.Command{
	Use:       "i3",
	Short:     "A brief description of your command",
	Long:      ``,
	ValidArgs: []string{"load", "disable"},
	Args:      cobra.ExactArgs(1),
        Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("i3 called")
	},
}
```

PASS `go run main.go i3` --> Error: accepts 1 arg(s), received 0
PASS `go run main.go i3 hello world` --> Error: accepts 1 arg(s), received 2
FAIL `go run main.go i3 chicken` --> i3 called
PASS `go run main.go i3 load ` --> i3 called

ValidArgs is being ignored because `cobra.OnlyValidArgs(cmd, args)` isn't called. When i pass in chicken, I want it to error because 'load' is the only valid arg.

Trying again with a custom Arg filter.
```
var i3Cmd = &cobra.Command{
	Use:       "i3",
	Short:     "A brief description of your command",
	Long:      ``,
	ValidArgs: []string{"load", "disable"},
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) != 1 {
			return errors.New("Requires exactly 1 arg.")
		}
		return cobra.OnlyValidArgs(cmd, args)
	},
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("i3 called")
	},
}
```

PASS `go run main.go i3 hello` --> Error: invalid argument "hello" for "dot i3"
PASS `go run main.go i3 load`  --> i3 called
PASS `go run main.go i3 hello world` --> Error: Requires exactly 1 arg
PASS `go run main.go i3` -->  Error: Requires exactly 1 arg.

The example above works... but it doesn't use `cobra.ExactArgs(1)`. I don't like re-inventing the wheel. How can I use both `OnlyValidArgs` and `ExactArgs` together? I tried:
```
var i3Cmd = &cobra.Command{
	Use:       "i3",
	Short:     "A brief description of your command",
	Long:      ``,
	ValidArgs: []string{"load"},
	Args: func(cmd *cobra.Command, args []string) error {
		err := cobra.ExactArgs(1)
		if err != nil {
			return err // <--- this is a compile error
		}	
		return cobra.OnlyValidArgs(cmd, args)
	},
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("i3 called")
	},
}
```
But `ExactArgs` returns `PositionalArgs`.. which looks to be a Type.. I'm not very familiar with Go, so this is confusing to me. I don't know how to combine the two. Any insight would be appreciated!

Also.. If I set `ValidArgs` and `OnlyValidArgs`, and an invalid command arg is provided, the user isn't notified of what the valid Args are. Is this the same feature request as #571? That seems like it would be very helpful.

related #284 
related #571 
related #346

**Comment by yoni386** [comment]
I would like to use OnlyValidArgs and MinimumNArgs(1)

**Comment by Antolius** [comment]
@patrick-motard for your exact use-case you can use `cobra.ExactValidArgs(1)` as a value for `ValidArgs`.

More general solution would be something like this:
```go
var cmd = &cobra.Command{
	Use:       "cmd",
	Short:     "General solution",
	ValidArgs: []string{"some", "acceptable", "values"},
	Args:      matchAll(cobra.MinimumNArgs(1), cobra.OnlyValidArgs),
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("cmd called")
	},
}

func matchAll(checks ...cobra.PositionalArgs) cobra.PositionalArgs {
	return func(cmd *cobra.Command, args []string) error {
		for _, check := range checks {
			if err := check(cmd, args); err != nil {
				return err
			}
		}
		return nil
	}
}
```
Here the `matchAll` function will effectively combine any number of `cobra.PositionalArgs` you pass to it, including the ones from `cobra` package and arbitrary ones you might have defined in your code.

If you think it would be a good idea to add this feature to the library itself I can submit a pull request.

**Comment by firelizzard18** [comment]
I would use the feature proposed by @Antolius

**Comment by umarcor** [comment]
@firelizzard18 you might want to try #841.

**Comment by firelizzard18** [comment]
@umarcor I want a function that composes multiple argument validators. Your MR does not provide this. Here is what I am using:

```golang
func args(fns ...cobra.PositionalArgs) cobra.PositionalArgs {
	return func(cmd *cobra.Command, args []string) error {
		for _, fn := range fns {
			if err := fn(cmd, args); err != nil {
				return err
			}
		}
		return nil
	}
}
```

**Comment by umarcor** [comment]
@firelizzard18, my bad. The reference should have been #896.

**Comment by firelizzard18** [comment]
@umarcor I +1'ed #896. Hopefully that will go somewhere...

**Comment by johnSchnake** [comment]
#841 is still active and other issues have mentioned it solves exactly this sort of problem. Leaving the issue open and unmarking as stale since the PR is in progress.

**Comment by marckhouzam** [comment]
The request and discussion here seem to be about composing multiple validators together.
This is now possible thanks to #896 and the new `MatchAll(pargs ...PositionalArgs)` function: https://github.com/spf13/cobra/blob/03c3eb73f8d00531742e232eb311f8cce0cd6225/args.go#L111-L112

**Comment by marckhouzam** [comment]
We now even have `ExactValidArgs(n int)` which seems to be exactly why this issue was opened: https://github.com/spf13/cobra/blob/03c3eb73f8d00531742e232eb311f8cce0cd6225/args.go#L89-L92

I'm going to go ahead and close this as fixed.

**Comment by umarcor** [comment]
@marckhouzam deprecating `ExactValidArgs` and `OnlyValidArgs` was proposed and approved in 2019, because that's an inconsistency in the implementation/codebase which is to be fixed in #841 (or #1643). Therefore, I do not recommend suggesting the usage of those functions. 

When #841 (or #1643) is merged, the default behaviour of `ExactArgs` + `ValidArgs` will be what the OP is asking.

Meanwhile, users can achieve the same result through `MatchAll`. However, that's meant for custom validators, not for combining the built-in ones. If we are going to suggest that as the official workaround for the incomplete implementation of `ValidArgs`, IMHO we'd better remove the feature from the codebase.
